cluster/images/etcd/migrate-if-needed.sh

```diff
@@ -34,22 +34,62 @@ if [ -z "${DATA_DIRECTORY:-}" ]; then
   exit 0
 fi
 
+ETCD="${ETCD:-/usr/local/bin/etcd}"
 ETCDCTL="${ETCDCTL:-/usr/local/bin/etcdctl}"
+ATTACHLEASE="${ATTACHLEASE:-/usr/local/bin/attachlease}"
 VERSION_FILE="version.txt"
 CURRENT_STORAGE='etcd2'
 if [ -e "${DATA_DIRECTORY}/${VERSION_FILE}" ]; then
   CURRENT_STORAGE="$(cat ${DATA_DIRECTORY}/${VERSION_FILE})"
 fi
 
+start_etcd() {
+  ETCD_PORT=18629
+  ETCD_PEER_PORT=18630
+  ${ETCD} --data-dir=${DATA_DIRECTORY} \
+    --listen-client-urls http://127.0.0.1:${ETCD_PORT} \
+    --advertise-client-urls http://127.0.0.1:${ETCD_PORT} \
+    --listen-peer-urls http://127.0.0.1:${ETCD_PEER_PORT} \
+    --initial-advertise-peer-urls http://127.0.01:${ETCD_PEER_PORT} \
+    1>>/dev/null 2>&1 &
+  ETCD_PID=$!
+  # Wait until etcd is up.
+  for i in $(seq 30); do
+    local out
+    if out=$(wget -q --timeout=1 http://127.0.0.1:${ETCD_PORT}/v2/machines 2> /dev/null); then
+      echo "Etcd on port ${ETCD_PORT} is up."
+      return 0
+    fi
+    sleep 0.5
+  done
+  echo "Timeout while waiting for etcd on port ${ETCD_PORT}"
+  return 1
+}
+
+stop_etcd() {
+  kill "${ETCD_PID-}" >/dev/null 2>&1 || :
+  wait "${ETCD_PID-}" >/dev/null 2>&1 || :
+}
+
 if [ "${CURRENT_STORAGE}" = "etcd2" -a "${TARGET_STORAGE}" = "etcd3" ]; then
   # If directory doesn't exist or is empty, this means that there aren't any
   # data for migration, which means we can skip this step.
   if [ -d "${DATA_DIRECTORY}" ]; then
     if [ "$(ls -A ${DATA_DIRECTORY})" ]; then
       echo "Performing etcd2 -> etcd3 migration"
-      # TODO: Pass a correct transformer to handle TTLs.
-      echo "ETCDCTL_API=3 ${ETCDCTL} migrate --data-dir=${DATA_DIRECTORY}"
       ETCDCTL_API=3 ${ETCDCTL} migrate --data-dir=${DATA_DIRECTORY}
+      echo "Attaching leases to TTL entries"
+      # Now attach lease to all keys.
+      # To do it, we temporarily start etcd on a random port (so that
+      # apiserver actually cannot access it).
+      start_etcd
+      # Create a lease and attach all keys to it.
+      ${ATTACHLEASE} \
+        --etcd-address http://127.0.0.1:${ETCD_PORT} \
+        --ttl-keys-prefix "${TTL_KEYS_DIRECTORY:-/registry/events}" \
+        --lease-duration 1m
```

Comment:
I think 10m is the value we use ordinarily? I guess you can make it a bit
higher since we're restarting everything. 15m?

On Wed, Aug 24, 2016 at 11:14 AM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In cluster/images/etcd/migrate-if-needed.sh
> <https://github.com/kubernetes/kubernetes/pull/31242#discussion_r76109794>
> :
>
> >        ETCDCTL_API=3 ${ETCDCTL} migrate --data-dir=${DATA_DIRECTORY}
> > +      echo "Attaching leases to TTL entries"
> > +      # Now attach lease to all keys.
> > +      # To do it, we temporarily start etcd on a random port (so that
> > +      # apiserver actually cannot access it).
> > +      start_etcd
> > +      # Create a lease and attach all keys to it.
> > +      ${ATTACHLEASE} \
> > +        --etcd-address http://127.0.0.1:${ETCD_PORT} \
> > +        --ttl-keys-prefix "${TTL_KEYS_DIRECTORY:-/registry/events}" \
> > +        --lease-duration 1m
>
> Actually, I set it to 1m for tests - good catch.
>
> Do you think that 10m is enough? Or should it be longer?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31242/files/1ff93b535adac0175574f6085e02d434121a8955#r76109794>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglppZzwDomNs8Txkh7IXnF9A3Sd0Wks5qjIoKgaJpZM4Jq7E_>
> .
>